File: 2022/03/solution.rb
#!ruby

input = (STDIN.tty? ? DATA : STDIN).readlines(chomp: true)
input = DATA.readlines(chomp: true) if input.size == 0

input.
  map do |rucksack|
    middle = rucksack.size / 2
    left = rucksack[...middle]
    right = rucksack[middle..]

    left.chars & right.chars
  end.
  flatten.
  map { _1.ord -  (_1 < 'a' ? 38 : 96) }.
  sum.
  then { puts _1 }

input.
  each_slice(3).
  map { _1.chars & _2.chars & _3.chars }.
  flatten.
  map { _1.ord -  (_1 < 'a' ? 38 : 96) }.
  sum.
  then { puts _1 }

__END__
vJrwpWtwJgWrhcsFMMfFFhFp
jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL
PmmdzqPrVvPwwTWBwg
wMqvLMZHhHMvwLHjbvcjnnSBnvTQFn
ttgJtRGJQctTZtZT
CrZsJsPPZsGzwwsLwLmpwMDw

File: 2022/03/solution.py
#!python

import sys
from functools import reduce

EXAMPLE = """
vJrwpWtwJgWrhcsFMMfFFhFp
jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL
PmmdzqPrVvPwwTWBwg
wMqvLMZHhHMvwLHjbvcjnnSBnvTQFn
ttgJtRGJQctTZtZT
CrZsJsPPZsGzwwsLwLmpwMDw
"""

input = EXAMPLE.splitlines()[1:] if sys.stdin.isatty() else sys.stdin.read().splitlines()
if len(input) == 0: input = input = EXAMPLE.splitlines()[1:]

def priority(char):
    if char < 'a':
        result = ord(char) - 38
    else:
        result = ord(char) - 96

    return result

result = [set(r[:len(r) // 2]) & set(r[len(r) // 2:]) for r in input]
result = map(lambda x: list(map(priority, x)), result)
result = sum(map(lambda x: sum(x), result))
print(result)

result = [input[i:i+3] for i in range(0, len(input), 3)]
result = map(lambda x: reduce(lambda a,b: set(a) & set(b), x), result)
result = map(lambda x: list(map(priority, x)), result)
result = sum(map(lambda x: sum(x), result))
print(result)


